### PR TITLE
Feat(cb2-5700): Amend PSV, HGV & TRL Tyres section

### PR DESCRIPTION
**Pull request by naathanbrown**
## Amend PSV, HGV & TRL Tyres section

Make the tyre section amendable including mock reference data.
[https://dvsa.atlassian.net/browse/CB2-5700](5700)

Worth noting the design has changed and been updated, the one in place represents what has been approved.

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 1]
```diff
@@ -41,24 +47,87 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
+
   get axles(): CustomFormArray {
     return this.form.get(['axles']) as CustomFormArray;
   }
 
+  getAxleTyres(i: number): CustomFormGroup {
+    return this.axles.get([i, 'tyres']) as CustomFormGroup;
+  }
+
+  getTyresRefData(tyre: Tyres, axleNumber: number) {
```
Comment (naathanbrown) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:90]: This kinda sucks, I'd really like to hear better ways around this if anyone has them

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 2]
```diff
@@ -41,24 +55,90 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
       case VehicleTypes.HGV:
-        return getTyresSection(VehicleTypes.HGV);
+        return tyresTemplateHgv;
       case VehicleTypes.TRL:
-        return getTyresSection(VehicleTypes.TRL);
+        return tyresTemplateTrl;
     }
   }
 
+  get isPsv(): boolean {
```
Comment (tom-bjss) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:66]: Functions like these feel like they should live in the techRecord service

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 3]
```diff
@@ -41,24 +55,90 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
+  }
+
+  getTyresRefData(tyre: Tyres, axleNumber: number) {
+    this.isError = false;
+    this.referenceDataService
+      .fetchReferenceData(ReferenceDataResourceType.Tyres, String(tyre.tyreCode))
+      .pipe(take(1))
+      .subscribe({
+        next: data => {
+          const newTyre = {
```
Comment (tom-bjss) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:105]: You can move this creation code of a Tyre interface into the interface model file, Ive seen this done in other typescript repos

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 4]
```diff
@@ -41,24 +55,90 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
+
+          const emptyTyre = {
+            tyreSize: null,
+            plyRating: null,
+            dataTrAxles: null,
+            speedCategorySymbol: tyre.speedCategorySymbol,
+            fitmentCode: tyre.fitmentCode,
+            tyreCode: tyre.tyreCode
+          };
+          this.vehicleTechRecord = cloneDeep(this.vehicleTechRecord);
```
Comment (tom-bjss) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:133]: duplicate of 117 - 119, you could even pass in the new tyre and do the nullcheck inside

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 5]
```diff
@@ -41,24 +55,90 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
+
+          this.vehicleTechRecord = cloneDeep(this.vehicleTechRecord);
+          this.vehicleTechRecord.axles.find(ax => ax.axleNumber === axleNumber)!.tyres = newTyre;
+          this.form.patchValue(this.vehicleTechRecord, { emitEvent: false });
+        },
+        error: _err => {
+          this.errorMessage = 'Cannot find data of this tyre';
+          this.isError = true;
+
+          const emptyTyre = {
```
Comment (tom-bjss) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:125]: Can also be moved into model interface file

**Review by tom-bjss — DISMISSED** [review]

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 6]
```diff
@@ -21,34 +36,115 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
+
   get axles(): CustomFormArray {
     return this.form.get(['axles']) as CustomFormArray;
   }
 
+  getAxleTyres(i: number): CustomFormGroup {
+    return this.axles.get([i, 'tyres']) as CustomFormGroup;
+  }
+
+  getTyresRefData(tyre: Tyres, axleNumber: number) {
```
Comment (aaron-downing) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:114]: Could we get some tests for these new methods please, I know they're likely to change in terms of the source of data but would be good to assure it works.

File: src/app/models/vehicle-tech-record.model.ts [hunk 7]
```diff
@@ -133,6 +134,10 @@ export enum SpeedCategorySymbol {
   Q = 'q'
 }
 
+export function speedCategorySymbolAsOptions(): MultiOptions {
```
Comment (aaron-downing) [inline src/app/models/vehicle-tech-record.model.ts:137]: We have a function for this in utils

File: src/app/forms/custom-sections/tyres/tyres.component.ts (hunk 6 above)
Comment (naathanbrown) [reply src/app/forms/custom-sections/tyres/tyres.component.ts:114]: I knew I had forgot to do something lol, yeah sure!

File: src/app/models/vehicle-tech-record.model.ts [hunk 8]
```diff
@@ -153,7 +162,24 @@ export interface Tyres {
...
+export class Tyre implements Tyres {
+  public tyreSize: string | null;
+  public speedCategorySymbol: SpeedCategorySymbol | null;
+  public fitmentCode: FitmentCode | null;
+  public dataTrAxles: number | null;
+  public plyRating: string | null;
+  public tyreCode: number | null;
+
+  constructor(public tyre: Tyres) {
+    (this.tyreSize = null),
```
Comment (aaron-downing) [inline src/app/models/vehicle-tech-record.model.ts:176]: Do all these need wrapping in brackets, or do they need ending in semicolons

File: src/app/forms/templates/trl/trl-tyres.template.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+          label: 'Axle',
+          value: '',
+          type: FormNodeTypes.GROUP,
+          children: [
+            {
+              name: 'axleNumber',
+              label: 'Axle Number',
+              type: FormNodeTypes.CONTROL,
+              validators: [
+                { name: ValidatorNames.Numeric },
```
Comment (aaron-downing) [inline src/app/forms/templates/trl/trl-tyres.template.ts:35]: If the axle number is editable and meant to be of type number can you use the NUMBER edit type please. if it's not editable are the validators necessary?

File: src/app/forms/templates/trl/trl-tyres.template.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+                  type: FormNodeTypes.CONTROL,
+                  validators: [{ name: ValidatorNames.Required }]
+                },
+                {
+                  name: 'dataTrAxles',
+                  label: 'Load index',
+                  value: '',
+                  type: FormNodeTypes.CONTROL,
+                  disabled: true,
+                  validators: [{ name: ValidatorNames.Numeric }, { name: ValidatorNames.Max, args: 999 }, { name: ValidatorNames.Min, args: 0 }]
```
Comment (aaron-downing) [inline src/app/forms/templates/trl/trl-tyres.template.ts:91]: Not editable?

Comment (naathanbrown) [reply src/app/forms/templates/trl/trl-tyres.template.ts:91]: Yeah, they are filled in by the ref data!

Comment (aaron-downing) [reply src/app/forms/templates/trl/trl-tyres.template.ts:91]: How would the validators handle if the ref data came back invalid? If the user has no control of what's in that field, does it make sense to have validators on it?

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 11]
```diff
@@ -21,34 +36,115 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
+            const newTyre = new Tyre(tyre);
+            newTyre.tyreSize = refTyre.tyreSize;
+            newTyre.plyRating = refTyre.plyRating;
+            newTyre.dataTrAxles = tyre.fitmentCode === FitmentCode.SINGLE ? Number(refTyre.loadIndexSingleLoad) : Number(refTyre.loadIndexTwinLoad);
+
+            this.addTyreToTechRecord(newTyre, axleNumber);
+          } catch {
+            this.errorMessage = 'Cannot find data of this tyre';
+            this.isError = true;
+            const newTyre = new Tyre(tyre);
```
Comment (aaron-downing) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:132]: could you declare the `newTyre` variable before your try block so you dont declare it twice (once in try, once in catch)?

File: src/app/forms/custom-sections/body/body.component.ts [hunk 12]
```diff
@@ -79,9 +73,7 @@ export class BodyComponent implements OnInit, OnChanges, OnDestroy {
   }
 
   get numberOptions(): MultiOptions {
-    return Array
-      .from(Array(10).keys())
-      .map(i => ({ value: i, label: `${i}` }));
+    return Array.from(Array(10).keys()).map(i => ({ value: i, label: `${i}` }));
```
Comment (gjulien-bjss) [inline src/app/forms/custom-sections/body/body.component.ts:76]: Not that it's super important, but I was staring at this function and thought if there was a way of  using `Object.fromEntries`. Turns out there isn't but this is the only improvement could come up with. `Array.from` can take a mapping function. So yeah just a suggestion
```suggestion
    return Array.from(Array(10).keys(), i => ({ value: i, label: `${i}` }));
```

Comment (naathanbrown) [reply src/app/forms/custom-sections/body/body.component.ts:76]: I've not actually changed this file, just my linter ran, so I'm gunna leave as is, sorry!

File: src/app/forms/custom-sections/tyres/tyres.component.ts [hunk 13]
```diff
@@ -21,34 +28,122 @@ export class TyresComponent implements OnInit, OnDestroy, OnChanges {
...
 
-  constructor(public dfs: DynamicFormService) {}
+  constructor(public dfs: DynamicFormService, private referenceDataService: ReferenceDataService) {}
 
   ngOnInit(): void {
-    const template = getTyresSection(this.vehicleTechRecord.vehicleType);
+    this.form = this.dfs.createForm(this.template, this.vehicleTechRecord) as CustomFormGroup;
+    this._formSubscription = this.form.cleanValueChanges.pipe(debounceTime(400)).subscribe(event => {
+      this.formChange.emit(event);
+    });
```
Comment (gjulien-bjss) [inline src/app/forms/custom-sections/tyres/tyres.component.ts:37]: Don't know if the formatting change was done on purpose 
```suggestion
    this._formSubscription = this.form.cleanValueChanges.pipe(debounceTime(400)).subscribe(event => this.formChange.emit(event));
```

**Review by gjulien-bjss — APPROVED** [review]

**Review by gjulien-bjss — APPROVED** [review]

**Review by aaron-downing — APPROVED** [review]

**Pull request merged**